# Catalog QC report

## Assemblies not found on NCBI

None

## Assemblies not found in UCSC list

None

## Assemblies with gene model URLs not found

- GCF_018416015.1
- GCF_002986195.1
- GCF_002986165.1

## Species and strain combinations with multiple taxonomy IDs

- Acanthamoeba castellanii strain Neff: 5755, 1257118
- Alphapapillomavirus 12: 10570, 990303
- Alphapapillomavirus 14: 120686, 333769
- Alphapapillomavirus 6: 10611, 333765
- Asfivirus haemorrhagiae: 443876, 443878, 10497
- Batrachochytrium dendrobatidis strain JEL423: 109871, 403673
- Betacoronavirus pandemicum: 227984, 2697049
- Betapapillomavirus 1: 333923, 889813
- Blumeria graminis: 1689686, 62690
- Candida albicans strain SC5314: 237561, 5476
- Candida tropicalis strain MYA-3404: 5482, 294747
- Cryptococcus gattii strain WM276: 367775, 37769
- Cryptococcus neoformans strain A5-35-17: 1230080, 5207
- Cryptococcus neoformans strain C23: 5207, 1295831
- Cryptococcus neoformans strain H99: 235443, 5207
- Cryptosporidium parvum: 353152, 5807
- Deltaretrovirus priTlym1: 11908, 194440
- Dependoparvovirus mammalian1: 82300, 256548
- Dependoparvovirus primate1: 202812, 202813, 10804, 85106, 57579
- Encephalitozoon intestinalis strain ATCC 50506: 876142, 58839
- Enterovirus alphacoxsackie: 150846, 156647
- Enterovirus betarhino: 44130, 12131
- Fusarium fujikuroi strain B14: 5127, 1215603
- Fusarium oxysporum: 396575, 1229664, 224912, 5507, 654392, 299031, 61374, 1940294, 330941, 61366
- Fusarium oxysporum strain Fo5176: 5507, 100902
- Fusarium vanettenii: 2747968, 660122
- Fusarium verticillioides strain 7600: 334819, 117187
- Gammapapillomavirus 11: 1195796, 1070409, 1070413
- Gammapapillomavirus 12: 746832, 909331
- Gammapapillomavirus 15: 1472342, 1070408
- Gammapapillomavirus 19: 1315259, 1315264
- Gemykibivirus humas1: 1519409, 1516081
- Glossina fuscipes: 7396, 201502
- Heligmosomoides polygyrus: 375939, 6339
- Hemileia vastatrix: 1291517, 203904
- Hepacivirus hominis: 1544901, 356114, 33745
- Meyerozyma guilliermondii strain ATCC 6260: 294746, 4929
- Microsporum canis strain CBS 113480: 63405, 554155
- Mucor lusitanicus strain CBS 277.49: 29924, 747725
- Neospora caninum strain Liverpool: 572307, 29176
- Norovirus norwalkense: 1246677, 1529918, 490039, 122928, 1529924, 1529909, 122929
- Orthoflavivirus denguei: 11069, 11070, 11053
- Orthomarburgvirus marburgense: 3052505, 448086
- Pegivirus columbiaense: 1729141, 1704090
- Plasmodium falciparum: 36329, 5833
- Plasmodium vinckei: 138297, 54757, 5860, 119398, 138298
- Puccinia striiformis strain 93-210: 168172, 27350
- Small anellovirus: 289367, 289366
- Torbevirus brisa: 2571075, 2571077, 2571078, 3116878
- Trichophyton rubrum strain CBS 289.86: 1215329, 5551
- Trichophyton rubrum strain CBS 735.88: 5551, 1215332
- Trypanosoma cruzi strain Dm28c: 85057, 5693, 1416333
- Vesivirus exanthema: 146073, 35612

## Assemblies without ploidy information

None

## Outbreak descendant taxonomy IDs not found in genomes data

None

## Outdated assembly accessions

- GCF_000002765.5 (current: GCF_000002765.6)
- GCF_018416015.1 (current: GCF_018416015.2)
- GCF_000857325.2 (current: GCF_000857325.3)
- GCF_000499545.1 (current: GCF_000499545.2)
- GCA_020800235.1 (current: GCA_020800235.2)
- GCF_023375885.1 (current: GCF_023375885.2)
- GCA_000237925.2 (current: GCA_000237925.5)

## Suppressed or retired genomes

- GCF_001021295.1 (status: suppressed)
- GCF_000002825.2 (status: suppressed)
- GCF_000002545.3 (status: suppressed)
- GCF_000002525.2 (status: suppressed)
- GCF_015227675.2 (status: suppressed)
- GCF_000364345.1 (status: suppressed)
- GCF_000005575.2 (status: suppressed)
- GCF_013339725.1 (status: suppressed)
- GCF_016920705.1 (status: suppressed)
- GCF_000149205.2 (status: suppressed)
- GCF_000457365.2 (status: suppressed)
- GCF_000226395.1 (status: suppressed)
- GCF_000271745.1 (status: suppressed)
- GCF_002775015.1 (status: suppressed)
- GCF_001189475.1 (status: suppressed)
- GCF_000151735.1 (status: suppressed)
- GCF_001876365.2 (status: suppressed)
- GCF_006496715.2 (status: suppressed)
- GCF_020744495.1 (status: suppressed)
- GCF_000371365.1 (status: suppressed)
- GCF_002892825.2 (status: suppressed)
- GCF_000208615.1 (status: suppressed)
- GCF_000003085.2 (status: suppressed)
- GCF_000209185.1 (status: suppressed)
- GCF_000001985.1 (status: suppressed)
- GCF_000003835.1 (status: suppressed)
- GCF_000315645.1 (status: suppressed)
- GCF_000230375.1 (status: suppressed)
- GCF_000151335.2 (status: suppressed)
- GCF_000769155.1 (status: suppressed)
- GCF_000184105.1 (status: suppressed)
- GCF_000149685.1 (status: suppressed)
- GCF_000182805.3 (status: suppressed)
- GCF_000182985.1 (status: suppressed)
- GCF_000723685.1 (status: suppressed)
- GCF_001015335.1 (status: suppressed)
- GCF_000209485.1 (status: suppressed)
- GCF_000258005.1 (status: suppressed)
- GCF_000709005.1 (status: suppressed)
- GCA_000469785.1 (status: suppressed)
- GCF_048819275.1 (status: suppressed)

## GenBank assemblies with paired RefSeq accessions

- GCA_009017415.1 (paired RefSeq: GCF_009017415.1)
- GCA_000857325.3 (paired RefSeq: GCF_000857325.3)
- GCA_004788295.1 (paired RefSeq: GCF_004788295.1)
- GCA_004786575.1 (paired RefSeq: GCF_004786575.1)
- GCA_004786635.1 (paired RefSeq: GCF_004786635.1)
- GCA_003108605.1 (paired RefSeq: GCF_003108605.1)
- GCA_008766755.1 (paired RefSeq: GCF_008766755.1)
- GCA_008766715.1 (paired RefSeq: GCF_008766715.1)
- GCA_004788235.1 (paired RefSeq: GCF_004788235.1)
- GCA_008776015.1 (paired RefSeq: GCF_008776015.1)
- GCA_000003225.1 (paired RefSeq: GCF_000003225.4)
- GCA_014883705.1 (paired RefSeq: GCF_014883705.1)
- GCA_014883685.1 (paired RefSeq: GCF_014883685.1)
- GCA_014883715.1 (paired RefSeq: GCF_014883715.1)
- GCA_014883695.1 (paired RefSeq: GCF_014883695.1)
- GCA_001629985.1 (paired RefSeq: GCF_001629985.1)
- GCA_000237925.2 (paired RefSeq: GCF_000237925.1)
- GCA_000002985.3 (paired RefSeq: GCF_000002985.6)
- GCA_010183535.1 (paired RefSeq: GCF_010183535.1)
- GCA_000507365.1 (paired RefSeq: GCF_000507365.1)
- GCA_000149245.3 (paired RefSeq: GCF_000149245.1)
- GCA_000011425.1 (paired RefSeq: GCF_000011425.1)
- GCA_000091045.1 (paired RefSeq: GCF_000091045.1)
- GCA_000091225.2 (paired RefSeq: GCF_000091225.2)
- GCA_000146465.1 (paired RefSeq: GCF_000146465.1)
- GCA_000002655.1 (paired RefSeq: GCF_000002655.1)
- GCA_000143185.2 (paired RefSeq: GCF_000143185.2)
- GCA_000149305.1 (paired RefSeq: GCF_000149305.1)
- GCA_000004985.1 (paired RefSeq: GCF_000004985.1)
- GCA_000149335.2 (paired RefSeq: GCF_000149335.2)
- GCA_000003525.2 (paired RefSeq: GCF_000003525.1)
- GCA_000006405.1 (paired RefSeq: GCF_000006405.1)
- GCA_000003855.2 (paired RefSeq: GCF_000003855.2)
- GCA_000002855.2 (paired RefSeq: GCF_000002855.4)
- GCA_000006335.3 (paired RefSeq: GCF_000006335.3)
- GCA_000002545.2 (paired RefSeq: GCF_000002545.3)
- GCA_000149205.2 (paired RefSeq: GCF_000149205.2)
- GCA_000006255.1 (paired RefSeq: GCF_000006255.1)
- GCA_000182965.3 (paired RefSeq: GCF_000182965.3)
- GCA_000185945.1 (paired RefSeq: GCF_000185945.1)
- GCA_000240135.3 (paired RefSeq: GCF_000240135.3)
- GCA_000184455.3 (paired RefSeq: GCF_000184455.2)
- GCA_000149555.1 (paired RefSeq: GCF_000149555.1)
- GCA_000303195.2 (paired RefSeq: GCF_000303195.2)
- GCA_000277815.3 (paired RefSeq: GCF_000277815.2)
- GCA_000280035.2 (paired RefSeq: GCF_000280035.1)
- GCA_000151355.1 (paired RefSeq: GCF_000151355.1)
- GCA_000149425.1 (paired RefSeq: GCF_000149425.1)
- GCA_000149615.1 (paired RefSeq: GCF_000149615.1)
- GCA_000151425.1 (paired RefSeq: GCF_000151425.1)
- GCA_000150115.1 (paired RefSeq: GCF_000150115.1)
- GCA_000313135.1 (paired RefSeq: GCF_000313135.1)
- GCA_000203795.2 (paired RefSeq: GCF_000203795.2)
- GCA_000260195.2 (paired RefSeq: GCF_000260195.1)
- GCA_000271585.1 (paired RefSeq: GCF_000271585.1)
- GCA_000151145.1 (paired RefSeq: GCF_000151145.1)
- GCA_000149925.1 (paired RefSeq: GCF_000149925.1)
- GCA_000204055.1 (paired RefSeq: GCF_000204055.1)
- GCA_000275845.1 (paired RefSeq: GCF_000275845.1)
- GCA_000182565.2 (paired RefSeq: GCF_000182565.1)
- GCA_000151505.1 (paired RefSeq: GCF_000151505.1)
- GCA_000150975.2 (paired RefSeq: GCF_000150975.2)
- GCA_000300595.1 (paired RefSeq: GCF_000300595.1)
- GCA_000250985.1 (paired RefSeq: GCF_000250985.1)
- GCA_000192795.1 (paired RefSeq: GCF_000192795.1)
- GCA_000231115.1 (paired RefSeq: GCF_000231115.1)
- GCA_000149585.1 (paired RefSeq: GCF_000149585.1)
- GCA_000313525.1 (paired RefSeq: GCF_000313525.1)
- GCA_000149955.2 (paired RefSeq: GCF_000149955.1)
- GCA_000149385.1 (paired RefSeq: GCF_000149385.1)
- GCA_000271745.2 (paired RefSeq: GCF_000271745.1)
- GCA_000209485.1 (paired RefSeq: GCF_000209485.1)
- GCA_000328475.2 (paired RefSeq: GCF_000328475.2)
- GCA_000497045.1 (paired RefSeq: GCF_000497045.1)
- GCA_000403515.1 (paired RefSeq: GCF_000403515.1)
- GCA_000417875.1 (paired RefSeq: GCF_000417875.1)
- GCA_000747765.1 (paired RefSeq: GCF_000747765.1)
- GCA_000732125.1 (paired RefSeq: GCF_000732125.1)
- GCA_000738915.1 (paired RefSeq: GCF_000738915.1)
- GCA_000769155.2 (paired RefSeq: GCF_000769155.1)
- GCA_000836335.2 (paired RefSeq: GCF_000836335.1)
- GCA_000835755.2 (paired RefSeq: GCF_000835755.1)
- GCA_001632365.1 (paired RefSeq: GCF_001632365.1)
- GCA_001642395.1 (paired RefSeq: GCF_001642395.1)
- GCA_001890705.1 (paired RefSeq: GCF_001890705.1)
- GCA_001890905.1 (paired RefSeq: GCF_001890905.1)
- GCA_001890125.1 (paired RefSeq: GCF_001890125.1)
- GCA_001983325.1 (paired RefSeq: GCF_001983325.1)
- GCA_002117355.1 (paired RefSeq: GCF_002117355.1)
- GCA_002708625.1 (paired RefSeq: GCF_002708625.1)
- GCA_002600585.1 (paired RefSeq: GCF_002600585.1)
- GCA_003054445.1 (paired RefSeq: GCF_003054445.1)
- GCA_002954075.1 (paired RefSeq: GCF_002954075.1)
- GCA_002847045.1 (paired RefSeq: GCF_002847045.1)
- GCA_003184595.1 (paired RefSeq: GCF_003184595.1)
- GCA_003344945.1 (paired RefSeq: GCF_003344945.1)
- GCA_003324165.2 (paired RefSeq: GCF_003324165.1)
- GCA_003813185.1 (paired RefSeq: GCF_003813185.1)
- GCA_005221285.1 (paired RefSeq: GCF_005221285.1)
- GCA_005498985.1 (paired RefSeq: GCF_005498985.1)
- GCA_006535985.1 (paired RefSeq: GCF_006535985.1)
- GCA_008403515.1 (paired RefSeq: GCF_008403515.1)
- GCA_010111755.1 (paired RefSeq: GCF_010111755.1)
- GCA_011745365.1 (paired RefSeq: GCF_011745365.1)
- GCA_013085055.1 (paired RefSeq: GCF_013085055.1)
- GCA_013396195.1 (paired RefSeq: GCF_013396195.1)
- GCA_013340325.1 (paired RefSeq: GCF_013340325.1)
- GCA_014117465.1 (paired RefSeq: GCF_014117465.1)
- GCA_018416015.2 (paired RefSeq: GCF_018416015.2)
- GCA_019609905.1 (paired RefSeq: GCF_019609905.1)
- GCA_019915245.1 (paired RefSeq: GCF_019915245.1)
- GCA_021901695.1 (paired RefSeq: GCF_021901695.1)
- GCA_021653875.1 (paired RefSeq: GCF_021653875.1)
- GCA_024244105.1 (paired RefSeq: GCF_024244105.1)
- GCA_024244095.1 (paired RefSeq: GCF_024244095.1)
- GCA_025433545.1 (paired RefSeq: GCF_025433545.1)
- GCA_026262505.1 (paired RefSeq: GCF_026262505.1)
- GCA_026914185.1 (paired RefSeq: GCF_026914185.1)
- GCA_026873545.1 (paired RefSeq: GCF_026873545.1)
- GCA_029851405.1 (paired RefSeq: GCF_029851405.1)
- GCA_036417295.1 (paired RefSeq: GCF_036417295.1)
- GCA_036630325.1 (paired RefSeq: GCF_036630325.1)
- GCA_900079805.1 (paired RefSeq: GCF_900079805.1)
- GCA_900067095.1 (paired RefSeq: GCF_900067095.1)
- GCA_900044065.1 (paired RefSeq: GCF_900044065.1)
- GCA_900007375.1 (paired RefSeq: GCF_900007375.1)

## Taxonomy tree

Assembly count mismatch: None

List of leaves without species in lineage: None

List of taxa with assembly count 0: None

List of taxonomic levels specified in parameters but absent in tree: None
