# Catalog Workflows QC report

## Newly added workflows (could be configured and set active)

- Allele-based Pathogen Identification (OTHER)
- Assembly decontamination VGP9  (OTHER)
- COVID-19: consensus construction (OTHER)
- Clinical Metaproteomics Data Interpretation (OTHER)
- Clinical Metaproteomics Discovery Workflow (OTHER)
- Clinical Metaproteomics Quantitation (OTHER)
- Clinical Metaproteomics Verification Workflow (OTHER)
- Create GRO and TOP complex files (OTHER)
- End-to-End Tissue Microarray Analysis (OTHER)
- Fragment-based virtual screening using rDock for docking and SuCOS for pose scoring (OTHER)
- Gene-based Pathogen Identification (OTHER)
- Generate Nx and Size plots for multiple assemblies (OTHER)
- Generate a Clinical Metaproteomics Database (OTHER)
- Genome Assembly from Hifi reads - VGP3 (OTHER)
- Genome Assembly from Hifi reads with HiC phasing - VGP4 (OTHER)
- Genome Assembly with Pacbio Hifi reads and Trio data for phasing - VGP5 (OTHER)
- K-mer profiling and reads statistics VGP1 (OTHER)
- MAPseq to ampvis2 (OTHER)
- MGnify amplicon summary tables (OTHER)
- MGnify's amplicon pipeline v5.0 (OTHER)
- MGnify's amplicon pipeline v5.0 - ITS (OTHER)
- MGnify's amplicon pipeline v5.0 - Quality control PE (OTHER)
- MGnify's amplicon pipeline v5.0 - Quality control SE (OTHER)
- MGnify's amplicon pipeline v5.0 - rRNA prediction (OTHER)
- MMGBSA calculations with GROMACS (OTHER)
- Mass spectrometry: GCMS with metaMS (OTHER)
- Mass spectrometry: LC-MS preprocessing with XCMS (OTHER)
- MetaProSIP OpenMS 2.8 (OTHER)
- Metagenome-Assembled Genomes (MAGs) generation (OTHER)
- Mitogenome Assembly VGP0 (OTHER)
- Molecular formula assignment and recalibration with MFAssignR package. (OTHER)
- Multiplex Tissue Microarray Analysis (OTHER)
- Nanopore Preprocessing (OTHER)
- Parallel Accession Download (OTHER)
- Pathogen Detection PathoGFAIR Samples Aggregation and Visualisation (OTHER)
- Post-Assembly Quality Control and Contamination Check for Bacterial Genomes (OTHER)
- PretextMap Generation from 1 or 2 haplotypes (OTHER)
- Purge duplicate contigs from a diploid assembly VGP6 (OTHER)
- Purging duplicates in one haplotype VGP6b (OTHER)
- QCxMS Spectra Prediction from SDF (OTHER)
- QIIME2 IIa: Denoising (sequence quality control) and feature table creation (single-end) (OTHER)
- QIIME2 IIb: Denoising (sequence quality control) and feature table creation (paired-end) (OTHER)
- QIIME2 Ia: multiplexed data (single-end) (OTHER)
- QIIME2 Ib: multiplexed data (paired-end) (OTHER)
- QIIME2 Ic: Demultiplexed data (single-end) (OTHER)
- QIIME2 Id: Demultiplexed data (paired-end) (OTHER)
- QIIME2 VI: Diversity metrics and estimations (OTHER)
- QIIME2-III-V-Phylogeny-Rarefaction-Taxonomic-Analysis (OTHER)
- RNA Velocity Analysis: Velocyto for 10X Data from Bundled Output (OTHER)
- RNA Velocity Analysis: Velocyto for 10X Data with Filtered Barcodes (OTHER)
- Scaffolding with Hi-C data VGP8 (OTHER)
- Scaffolding-BioNano-VGP7 (OTHER)
- Segmentation and counting of cell nuclei in fluorescence microscopy images (OTHER)
- Single-Cell RNA-seq Analysis: Scanpy Preprocessing and Clustering (OTHER)
- Taxonomic abundance summary tables for a specified taxonomic rank (OTHER)
- Taxonomy Profiling and Visualization with Krona (OTHER)
- dada2 amplicon analysis pipeline - for paired end data (OTHER)
- dcTMD calculations with GROMACS (OTHER)
- kmer-profiling-hifi-trio-VGP2 (OTHER)
- sra_manifest_to_concatenated_fastqs_parallel (OTHER)

## Active workflows with version not on Dockstore

None

## Workflows not using newest IWC version (newer not on Dockstore yet)

None

## Workflows with unknown category and one valid category (kept)

- #workflow/github.com/iwc-workflows/fastq-to-matrix-10x/scrna-seq-fastq-to-matrix-10x-cellplex (categories: OTHER, TRANSCRIPTOMICS)
- #workflow/github.com/iwc-workflows/fastq-to-matrix-10x/scrna-seq-fastq-to-matrix-10x-v3 (categories: OTHER, TRANSCRIPTOMICS)
- #workflow/github.com/iwc-workflows/sars-cov-2-pe-illumina-artic-variant-calling/COVID-19-PE-ARTIC-ILLUMINA (categories: OTHER, VARIANT_CALLING)
- #workflow/github.com/iwc-workflows/sars-cov-2-pe-illumina-wgs-variant-calling/COVID-19-PE-WGS-ILLUMINA (categories: OTHER, VARIANT_CALLING)
- #workflow/github.com/iwc-workflows/sars-cov-2-se-illumina-wgs-variant-calling/COVID-19-SE-WGS-ILLUMINA (categories: OTHER, VARIANT_CALLING)

## Workflows with only unknown category (excluded)

None

## Workflows with multiple valid categories

- #workflow/github.com/iwc-workflows/generic-non-segmented-viral-variant-calling/main (categories: CONSENSUS_SEQUENCES, VARIANT_CALLING)

## Parameter changes by workflow

> **Note:** New parameters are added to workflows.yml on each run. On subsequent runs, they will no longer appear as 'new' even if stale parameters haven't been addressed yet. If you see both stale and new params for a workflow, commit or save this report before re-running so you don't lose the pairing info (useful for identifying renames).

None
